cpp/src/arrow/util/hash.h

```diff
@@ -39,6 +39,9 @@ static constexpr double kMaxHashTableLoad = 0.5;
 
 namespace internal {
 
+// TODO this ugliness should be rewritten as an inline function with
+// a callable argument.
```

Comment:
Agreed; this refactor should be accompanied by additional benchmarks to ensure we do not introduce performance regressions